### ISSUE TITLE
PyInstaller does not work with cx_Oracle and the Oracle Instant Client

### ISSUE THREAD

**Issue by SarahLuca** [pyinstaller/pyinstaller#5572]
I am trying to freeze a python application that uses cx_oracle and the oracle instant client to connect to an oracle database. I am working on Windows 10. My application works fine before I package it. But after packaging, the following error occurs when I try to run the executable file:

**cx_Oracle.DatabaseError: DPI-1072: the Oracle Client library version is unsupported**

This error occurs in the line ```cx_Oracle.init_oracle_client(lib_dir=r"instantclient_19_9")```. According to the ODPI log, the Oracle Instant Client is loaded successfully, but while validating, the error occurs.

This error can be replicated with the following **toy example**:

There needs to be no Oracle database present. The Oracle Instant Client needs to be installed into the folder where the following file is located.

```
import cx_Oracle

if __name__ == "__main__":
    cx_Oracle.init_oracle_client(lib_dir=r"instantclient_19_9")
```

**Comment by bwoodsend** [comment]
What's the full traceback?

**Comment by rokm** [comment]
It seems that Oracle Client library directory needs to be explicitly added to `PATH`, for whatever reason... I.e.,

```python
# build with:
#  pyinstaller program.py --add-data "instantclient_19_10;instantclient_19_10" --clean --noconfirm

import sys
import os

instaclient_dir = 'instantclient_19_10'
if getattr(sys, 'frozen', False):
    instaclient_dir = os.path.join(sys._MEIPASS, instaclient_dir)
    # Add Oracle Client library directory to PATH
    os.environ['PATH'] = instaclient_dir + os.path.pathsep + os.environ['PATH']

import cx_Oracle
if __name__ == "__main__":
    cx_Oracle.init_oracle_client(lib_dir=instaclient_dir)
    print("Hello world!")
```

**Comment by SarahLuca** [comment]
> It seems that Oracle Client library directory needs to be explicitly added to `PATH`, for whatever reason... I.e.,
> 
> ```python
> # build with:
> #  pyinstaller program.py --add-data "instantclient_19_10;instantclient_19_10" --clean --noconfirm
> 
> import sys
> import os
> 
> instantclient_dir = 'instantclient_19_10'
> if getattr(sys, 'frozen', False):
>     instaclient_dir = os.path.join(sys._MEIPASS, instantclient_dir)
>     # Add Oracle Client library directory to PATH
>     os.environ['PATH'] = instantclient_dir + os.path.pathsep + os.environ['PATH']
> 
> import cx_Oracle
> if __name__ == "__main__":
>     cx_Oracle.init_oracle_client(lib_dir=instantclient_dir)
>     print("Hello world!")
> ```

**Comment by cjbj** [comment]
Are you sure you didn't need a directory path before the "instantclient_19_9" in the init_oracle_client call?

**Comment by SarahLuca** [comment]
It works using the relative path. The folder instantclient_19_9 is located in the same folder as the executable is. The executable is run from this folder.

**Comment by cjbj** [comment]
@SarahLuca what works?  Setting PATH and calling init_oracle_client()?  Did you try just setting PATH?  I'd like to know exactly.  This will help us see if we can make improvements.

**Comment by mmarquet** [comment]
Even tho the issue is closed, I can confirm I reproduced the exact same behavior and adding the instantclient folder to the path for the frozen exe fixed the issue somehow.

**Comment by DylanTheThird** [comment]
I think i solved the issue. I'll write here and in couple other places so other will not waste their lives and time for this.
Somehow, somewhere, I found information how cx_oracle sets up path to instantclient. While it should be obvious to use init_oracle_client, where you provide the correct path, it does not work on win. Why? Someone had stupid idea that on win, everybody is putting oracle client in one place. And so, in import configuration, there is line
if sys.os == win
c:\oracle\instantclient
So after 2 days of guts pulling with pyinstaller and py2exe i decided to humor them ad put my instant client in this path. And yeah, it works.
Mind you, not oracle\instantclient directly but just unpack your client in whatever folder is named, i got instantclient_23_3 and put in in c:\oracle. Then pyinstaller will work without additional steps.
Also, I made button to provide path to instant client in program.

**Comment by cjbj** [comment]
@DylanTheThird Thanks for sharing.

For reference, the Instant Client load path is described in the manual https://cx-oracle.readthedocs.io/en/latest/user_guide/initialization.html

You can trace what is being looked for and being loaded by setting the environment variable DPI_DEBUG_LEVEL=64 (this is also shown in the above link)

**Comment by DylanTheThird** [comment]
That description is wrong. If it was right, there wouldn't be problems with it.
In my case and couple other cases on the net, the problem is in the import itself. Just make a python file, for example(I'm using PyCharm)
```
import cx_Oracle

print('this suppose to work on other OS but for some unknown reason, it will probably not work on win')
```

pack it into exe using pyinstaller
run
and it might give error that cx_oracle module is not found.

looks like after instantclient is in oracle path, then whatever I try it works. But in case someone still have problem, try that entire
```
try:
    if sys.platform.startswith("darwin"):
```
code from https://cx-oracle.readthedocs.io/en/latest/user_guide/initialization.html
it might help.

**Comment by cjbj** [comment]
@DylanTheThird I'd love to help more.  Can you set the environment variable (in a non working scenario) and share the trace output?

**Comment by DylanTheThird** [comment]
ok, i'll try to break it again and set that variable.

**Comment by DylanTheThird** [comment]
I got nothing.
When running through pycharm or the exe file I get:
ODPI [14312] 2021-12-28 12:56:03.642: ODPI-C 4.3.0
ODPI [14312] 2021-12-28 12:56:03.642: debugging messages initialized at level 64
test

I removed the oracle from C folder, I changed name of the oracle folders and its still working.
I tried on another pc, where it should be raw and I get error about missing module but all I get is
Traceback (most recent call last):
  File "testOracle.py", line 1, in <module>
ModuleNotFoundError: No module named 'cx_Oracle'
[5912] Failed to execute script testOracle
when running exe file.
